src/Npgsql/ConnectorSource.cs

```diff
@@ -0,0 +1,86 @@
+﻿using Npgsql.Util;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Transactions;
+
+namespace Npgsql
+{
+    abstract class ConnectorSource
+    {
+        internal NpgsqlConnectionStringBuilder Settings { get; }
+
+        /// <summary>
+        /// Contains the connection string returned to the user from <see cref="NpgsqlConnection.ConnectionString"/>
+        /// after the connection has been opened. Does not contain the password unless Persist Security Info=true.
+        /// </summary>
+        internal string UserFacingConnectionString { get; }
+
+        // Note that while the dictionary is protected by locking, we assume that the lists it contains don't need to be
+        // (i.e. access to connectors of a specific transaction won't be concurrent)
+        readonly Dictionary<Transaction, List<NpgsqlConnector>> _pendingEnlistedConnectors
+            = new();
+
+        internal abstract (int Total, int Idle, int Busy) Statistics { get; }
+
+        internal ConnectorSource(NpgsqlConnectionStringBuilder settings, string connString)
+        {
+            Settings = settings;
+
+            UserFacingConnectionString = settings.PersistSecurityInfo
+                ? connString
+                : settings.ToStringWithoutPassword();
+        }
+
+        internal abstract ValueTask<NpgsqlConnector> Get(
+            NpgsqlConnection conn, NpgsqlTimeout timeout, bool async, CancellationToken cancellationToken);
+
+        internal abstract void Return(NpgsqlConnector connector);
+
+        internal abstract void Clear();
+
+        #region Pending Enlisted Connections
+
+        internal virtual void AddPendingEnlistedConnector(NpgsqlConnector connector, Transaction transaction)
+        {
+            lock (_pendingEnlistedConnectors)
+            {
+                if (!_pendingEnlistedConnectors.TryGetValue(transaction, out var list))
+                    list = _pendingEnlistedConnectors[transaction] = new List<NpgsqlConnector>();
+                list.Add(connector);
+            }
+        }
+
+        internal virtual void TryRemovePendingEnlistedConnector(NpgsqlConnector connector, Transaction transaction)
+        {
+            lock (_pendingEnlistedConnectors)
+            {
+                if (!_pendingEnlistedConnectors.TryGetValue(transaction, out var list))
+                    return;
+                list.Remove(connector);
+                if (list.Count == 0)
+                    _pendingEnlistedConnectors.Remove(transaction);
+            }
+        }
+
+        internal virtual bool TryRentEnlistedPending(Transaction transaction, [NotNullWhen(true)] out NpgsqlConnector? connector)
```

Comment:
I think there's a problem here... Consider the following scenario:

* User starts TransactionScope
* Open connection with TargetSessionAttributes with Secondary, closes. Connection is now enlisted pending.
* Open connection with TargetSessionAttributes with Primary.

If we return the previous pending transaction, we violate the TargetSessionAttributes. So we need to check whether the pending connection is compatible with the given TargetSessionAttributes. If not, TryRentEnlistedPending can return false, and the open trigger escalation to a distributed transaction, which will (currently) fail on .NET Core.